File: static_files/assignments/HW2_files/SRC/MATLAB/gen_fx_get_svm.m
unction [out , mdl] = gen_fx_get_svm(grp, I, rate, rep)
% gen_fx_get_svm - Trains and evaluates an SVM classifier over multiple repetitions.
%
% Syntax:
%   [out, mdl] = gen_fx_get_svm(grp, I, rate, rep)
%
% Inputs:
%   grp  - A vector of class labels for each sample (Nx1)
%   I    - Feature matrix (NxD), where N is the number of samples and D is the number of features
%   rate - Proportion of data to use for training in each repetition (e.g., 0.7 for 70%)
%   rep  - Number of repetitions to run the evaluation
%
% Outputs:
%   out - Struct with fields:
%         pt  - Vector of classification accuracies for each repetition
%         C   - Confusion matrices for each repetition
%         tu  - Vector of per-class accuracies for each repetition
%         model - The SVM model from the last repetition
%   mdl - Same as out.model (for backward compatibility or convenience)
%
% This function repeatedly splits the dataset into training and testing subsets,
% trains an SVM classifier, evaluates its performance, and accumulates the results.

Pt = [];  % Accuracy scores for each repetition
Tu = [];  % Per-class accuracies for each repetition

for cnt = 1:rep
    % Split the data into training and testing using equal class distribution
    [test, train] = gen_fx_get_equal_part(grp, rate);

    % Train and test the SVM classifier
    [cls , mdl] = gen_fx_MC_SVM(I(test,:), I(train,:), grp(train));

    % Calculate accuracy for this repetition
    Pt = [Pt; sum(cls == grp(test)) / sum(test)];

    % Compute the confusion matrix
    Ct(:,:,cnt) = confusionmat(cls, grp(test));

    % Compute per-class accuracy for this repetition
    Tu = [Tu diag(Ct(:,:,cnt)) ./ (sum(Ct(:,:,cnt)))'];
end

% Store results in the output structure
out.C = Ct;
out.pt = Pt;
out.tu = Tu;
out.model = mdl;
end

File: static_files/assignments/HW2_files/SRC/Python/gen_fx_get_svm.py

import numpy as np
from sklearn.metrics import confusion_matrix
from gen_fx_get_equal_part import gen_fx_get_equal_part
from gen_fx_MC_SVM import gen_fx_MC_SVM

def gen_fx_get_svm(grp, I, rate, rep):
    """
    Trains and evaluates an SVM classifier over multiple repetitions.

    Parameters:
        grp (np.ndarray): Class labels (N,)
        I (np.ndarray): Feature matrix (N, D)
        rate (float): Proportion of training data per class
        rep (int): Number of repetitions

    Returns:
        out (dict): Contains accuracy, confusion matrices, per-class accuracy, and last model
        mdl: Last trained model
    """
    grp = np.array(grp)
    Pt = []  # Accuracy scores
    Tu = []  # Per-class accuracy
    Ct = []  # Confusion matrices

    for _ in range(rep):
        test, train = gen_fx_get_equal_part(grp, rate)

        cls, mdl = gen_fx_MC_SVM(I[test], I[train], grp[train])

        accuracy = np.sum(cls == grp[test]) / np.sum(test)
        Pt.append(accuracy)

        conf = confusion_matrix(grp[test], cls)
        Ct.append(conf)

        per_class_acc = np.diag(conf) / np.sum(conf, axis=1)
        Tu.append(per_class_acc)

    out = {
        "pt": np.array(Pt),
        "C": np.array(Ct),
        "tu": np.array(Tu),
        "model": mdl
    }

    return out, mdl
